File: src/fshafas/HafasClient.fs
namespace FsHafas.Api

#if FABLE_COMPILER
open Fable.Core
open Fable.Core.JsInterop
#endif

open FsHafas.Client

/// <summary>JS promise based interface corresponding to hafas-client</summary>
type HafasClient(profile: FsHafas.Client.Profile) =
    let client =
        new FsHafas.Api.HafasAsyncClient(profile :?> FsHafas.Endpoint.Profile)

#if FABLE_COMPILER
    [<Emit("typeof $1")>]
    let jsTypeof (_: obj) : string = jsNative

    [<Emit("$1.type")>]
    let jsTypeField (o: obj) : string = jsNative

    // get cases at runtime
    let makeCaseOfU4 (v: U4<string, Station, Stop, Location>) : U4<string, Station, Stop, Location> =
        if jsTypeof v = "string" then
            U4.Case1(unbox<string> v)
        else if (jsTypeField v) = "station" then
            let s = unbox<Station> v
            U4.Case2 { Default.Station with id = s.id }
        else if (jsTypeField v) = "stop" then
            let s = unbox<Stop> v
            U4.Case3 { Default.Stop with id = s.id }
        else if (jsTypeField v) = "location" then
            let l = unbox<Location> v

            U4.Case4
                { Default.Location with
                    id = l.id
                    poi = l.poi
                    address = l.address
                    latitude = l.latitude
                    longitude = l.longitude }
        else
            raise (System.ArgumentException("string|Station|Stop|Location expected"))

    // get cases at runtime
    let makeCaseOfU2StringStop (v: U2<string, Stop>) : U2<string, Stop> =
        if jsTypeof v = "string" then
            U2.Case1(unbox<string> v)
        else if (jsTypeField v) = "stop" then
            let s = unbox<Stop> v
            U2.Case2 { Default.Stop with id = s.id }
        else
            raise (System.ArgumentException("string|Stop expected"))

    let validateString (v: string) =
        if jsTypeof v <> "string"
           || (unbox<string> v).Length = 0 then
            raise (System.ArgumentException("string expected"))

    // make Location at runtime
    let makeLocation (v: Location) : Location =
        if (jsTypeField v) = "location" then
            let l = unbox<Location> v

            { Default.Location with
                id = l.id
                latitude = l.latitude
                longitude = l.longitude
                address = l.address }
        else
            raise (System.ArgumentException("Location expected"))

    let validateBoundingBox (v: BoundingBox) =
        if jsTypeof v.north <> "number"
           || jsTypeof v.west <> "number"
           || jsTypeof v.south <> "number"
           || jsTypeof v.east <> "number" then
            raise (System.ArgumentException("BoundingBox expected"))
#endif

    interface FsHafas.Client.HafasClient with

        member __.journeys
            (from: U4<string, Station, Stop, Location>)
            (``to``: U4<string, Station, Stop, Location>)
            (opt: JourneysOptions option)
            =
#if FABLE_COMPILER
            client.AsyncJourneys(makeCaseOfU4 from) (makeCaseOfU4 ``to``) opt
            |> Async.StartAsPromise
#else
            client.AsyncJourneys from ``to`` opt
#endif

        member __.refreshJourney (refreshToken: string) (opt: RefreshJourneyOptions option) =
#if FABLE_COMPILER
            validateString refreshToken

            client.AsyncRefreshJourney refreshToken opt
            |> Async.StartAsPromise
#else
            client.AsyncRefreshJourney refreshToken opt
#endif

        member __.trip (id: string) (name: string) (opt: TripOptions option) =
#if FABLE_COMPILER
            validateString id

            client.AsyncTrip id name opt
            |> Async.StartAsPromise
#else
            client.AsyncTrip id name opt
#endif

        member __.departures (id: U4<string, Station, Stop, Location>) (opt: DeparturesArrivalsOptions option) =
#if FABLE_COMPILER
            client.AsyncDepartures(makeCaseOfU4 id) opt
            |> Async.StartAsPromise
#else
            client.AsyncDepartures id opt
#endif

        member __.arrivals (id: U4<string, Station, Stop, Location>) (opt: DeparturesArrivalsOptions option) =
#if FABLE_COMPILER
            client.AsyncArrivals(makeCaseOfU4 id) opt
            |> Async.StartAsPromise
#else
            client.AsyncArrivals id opt
#endif

#if FABLE_COMPILER
        member __.journeysFromTrip
            (tripId: string)
            (prevStopOver: StopOver)
            (``to``: U4<string, Station, Stop, Location>)
            (opt: JourneysFromTripOptions option)
            : JS.Promise<array<Journey>> =
            client.AsyncJourneysFromTrip tripId prevStopOver (makeCaseOfU4 ``to``) opt
            |> Async.StartAsPromise
#else
        member __.journeysFromTrip
            (tripId: string)
            (prevStopOver: StopOver)
            (``to``: U4<string, Station, Stop, Location>)
            (opt: JourneysFromTripOptions option)
            : Async<array<Journey>> =
            client.AsyncJourneysFromTrip tripId prevStopOver ``to`` opt
#endif

        member __.locations (name: string) (opt: LocationsOptions option) =
#if FABLE_COMPILER
            validateString name

            client.AsyncLocations name opt
            |> Async.StartAsPromise
#else
            client.AsyncLocations name opt
#endif

        member __.stop (stop: U2<string, Stop>) (opt: StopOptions option) =
#if FABLE_COMPILER
            client.AsyncStop(makeCaseOfU2StringStop stop) opt
            |> Async.StartAsPromise
#else
            client.AsyncStop stop opt
#endif

        member __.nearby (location: Location) (opt: NearByOptions option) =
#if FABLE_COMPILER
            client.AsyncNearby(makeLocation location) opt
            |> Async.StartAsPromise
#else
            client.AsyncNearby location opt
#endif

        member __.reachableFrom (location: Location) (opt: ReachableFromOptions option) =
#if FABLE_COMPILER
            client.AsyncReachableFrom(makeLocation location) opt
            |> Async.StartAsPromise
#else
            client.AsyncReachableFrom location opt
#endif

        member __.radar (x: BoundingBox) (opt: RadarOptions option) =
#if FABLE_COMPILER
            validateBoundingBox x
            client.AsyncRadar x opt |> Async.StartAsPromise
#else
            client.AsyncRadar x opt
#endif

        member __.tripsByName (name: string) (opt: TripsByNameOptions option) =
#if FABLE_COMPILER
            validateString name

            client.AsyncTripsByName name opt
            |> Async.StartAsPromise
#else
            client.AsyncTripsByName name opt
#endif

        member __.remarks(opt: RemarksOptions option) =
#if FABLE_COMPILER
            client.AsyncRemarks opt |> Async.StartAsPromise
#else
            client.AsyncRemarks opt
#endif

        member __.lines (query: string) (opt: LinesOptions option) =
#if FABLE_COMPILER
            validateString query

            client.AsyncLines query opt
            |> Async.StartAsPromise
#else
            client.AsyncLines query opt
#endif

        member __.serverInfo(opt: ServerOptions option) =
#if FABLE_COMPILER
            client.AsyncServerInfo opt |> Async.StartAsPromise
#else
            client.AsyncServerInfo opt
#endif

File: src/fshafas.python.package/fshafas/hafas_client.py
from __future__ import annotations
from typing import (Optional, List, Any, Tuple, TypeVar,
                    Tuple, Awaitable, Union, Callable, Dict)
from .fable_modules.fable_library.array_ import (map)
from .fable_modules.fable_library.task import TaskCompletionSource
from .fable_modules.fable_library.async_builder import (singleton, Async)
from .fable_modules.fable_library.util import IDisposable
from .fable_modules.fable_library.async_ import (
    start_with_continuations, start_as_task, default_cancellation_token)
from .fable_modules.fable_library.util import (
    create_atom, ignore, structural_hash, string_hash)
from .fable_modules.fs_hafas_api_python.types_hafas_client import (Profile, IndexMap_2, Station, Stop, Location, StationStopLocation, ProductType, JourneysOptions, Journeys, RefreshJourneyOptions, Journey, StopOver, JourneysFromTripOptions, TripOptions, Trip, DeparturesArrivalsOptions,
                                                               TripsWithRealtimeData, Departures, Arrivals, LocationsOptions, StopOptions, Location, NearByOptions, ReachableFromOptions, Duration, BoundingBox, RadarOptions, Radar, TripsByNameOptions, RemarksOptions, Warning, LinesOptions, Line, ServerOptions, ServerInfo)
from .fable_modules.fs_hafas_python.hafas_async_client import (HafasAsyncClient_productsOfMode, HafasAsyncClient)
from .fable_modules.db_vendo_python.db_vendo_async_client import (DbVendoAsyncClient)
from .fable_modules.fs_hafas_python.context import (
    Profile as InternalProfile)
from .fable_modules.fs_hafas_api_python.defaults import (
    LocationsOptions as Default_LocationsOptions, JourneysOptions as Default_JourneysOptions, Location as Default_Location, DeparturesArrivalsOptions as Default_DeparturesArrivalsOptions)

# todo: implement all methods of HafasAsyncClient

def _isDbVendoProfile(profile: Any):
    return profile.endpoint == "https://app.vendo.noncd.db.de"
    
def _checkProfileType(profile: Any):
    # check, if profile is from fshafas.fable_modules.fs_hafas_profiles_python classes
    if _isDbVendoProfile(profile): 
        return
    elif not isinstance(profile, InternalProfile):
        raise TypeError(
            "argument profile: type from fshafas.fable_modules.fs_hafas_profiles_python expected ")

class HafasClient(IDisposable):
    def __init__(self, profile: Profile) -> None:
        _checkProfileType(profile)

        self.profile = profile
        if _isDbVendoProfile(profile):
            self.asyncClient = DbVendoAsyncClient()
        else:
            self.asyncClient = HafasAsyncClient(self.profile)

    async def _journeys(self, _from: Union[str, Station, Stop, Location], _to: Union[str, Station, Stop, Location], opt: Optional[JourneysOptions] = None) -> Journeys:
        def generate() -> Async[Journeys]:
            def bind(_arg1: Journeys) -> Async[Journeys]:
                return singleton.Return(_arg1)

            return singleton.Bind(self.asyncClient.AsyncJourneys(_from, _to, opt), bind)

        return await start_as_task(singleton.Delay(generate))

    async def journeys(self, _from: Union[str, Station, Stop, Location], _to: Union[str, Station, Stop, Location], opt: Optional[JourneysOptions] = None) -> Journeys:
        """Get journeys between locations."""
        if opt is None:
            opt = Default_JourneysOptions

        if (isinstance(_from, str) and not _from.isdigit()):
            from_locs= await self.locations(_from, Default_LocationsOptions)
            if (len(from_locs) > 0 and from_locs[0].fields[0].type == "stop"):
                _from= from_locs[0].fields[0].id

        if (isinstance(_to, str) and not _to.isdigit()):
            to_locs= await self.locations(_to, Default_LocationsOptions)
            if (len(to_locs) > 0 and to_locs[0].fields[0].type == "stop"):
                _to= to_locs[0].fields[0].id

        return await self._journeys(_from, _to, opt)

    async def bestprices(self, _from: str,  _to: str, opt: Optional[JourneysOptions]=None) -> Journeys:
        """Get best prices for journeys."""
        if (opt is not None and not isinstance(opt, JourneysOptions)):
            raise TypeError(
                "argument opt: type JourneysOptions expected")

        if (not _from.isdigit()):
            raise TypeError(
                "argument _from: station id expected")

        if (not _to.isdigit()):
            raise TypeError(
                "argument _to: station id expected")

        def generate() -> Async[Journeys]:
            def bind(_arg1: Journeys) -> Async[Journeys]:
                return singleton.Return(_arg1)

            return singleton.Bind(self.asyncClient.AsyncBestPrices( _from, _to, opt), bind)

        return await start_as_task(singleton.Delay(generate))

    async def departures(self, name: Union[str, Station, Stop, Location], opt: Optional[DeparturesArrivalsOptions]=None) -> Departures:
        """Query the next departures at a station."""
        if (opt is not None and not isinstance(opt, DeparturesArrivalsOptions)):
            raise TypeError(
                "argument opt: type DeparturesArrivalsOptions expected")

        def generate() -> Async[Departures]:
            def bind(_arg1: Departures) -> Async[Departures]:
                return singleton.Return(_arg1)

            return singleton.Bind(self.asyncClient.AsyncDepartures(name, opt), bind)

        return await start_as_task(singleton.Delay(generate))

    async def arrivals(self, name: Union[str, Station, Stop, Location], opt: Optional[DeparturesArrivalsOptions]=None) -> Arrivals:
        """Query the next arrivals at a station."""
        if (opt is not None and not isinstance(opt, DeparturesArrivalsOptions)):
            raise TypeError(
                "argument opt: type DeparturesArrivalsOptions expected")

        def generate() -> Async[Arrivals]:
            def bind(_arg1: Arrivals) -> Async[Arrivals]:
                return singleton.Return(_arg1)

            return singleton.Bind(self.asyncClient.AsyncArrivals(name, opt), bind)

        return await start_as_task(singleton.Delay(generate))

    async def locations(self, name: str, opt: Optional[LocationsOptions]=None) -> List[StationStopLocation]:
        """Find stations, POIs and addresses."""
        if opt is None:
            opt = Default_LocationsOptions

        if (not isinstance(name, str)):
            raise TypeError("argument name: type string expected")

        if (opt is not None and not isinstance(opt, LocationsOptions)):
            raise TypeError("argument opt: type LocationsOptions expected")

        def generate() -> Async[List[Any]]:
            def bind(_arg1: List[Any]) -> Async[List[Any]]:
                return singleton.Return(_arg1)

            return singleton.Bind(self.asyncClient.AsyncLocations(name, opt), bind)

        return await start_as_task(singleton.Delay(generate))

    async def nearby(self, l: Location, opt: Optional[NearByOptions]=None) -> List[StationStopLocation]:
        """Show stations & POIs around."""
        if (not isinstance(l, Location)):
            raise TypeError("argument l: type Location expected")

        if (opt is not None and not isinstance(opt, NearByOptions)):
            raise TypeError("argument opt: type NearByOptions expected")

        def generate() -> Async[List[Any]]:
            def bind(_arg1: List[Any]) -> Async[List[Any]]:
                return singleton.Return(_arg1)

            return singleton.Bind(self.asyncClient.AsyncNearby(l, opt), bind)

        return await start_as_task(singleton.Delay(generate))

    async def radar(self, rect: BoundingBox, opt: Optional[RadarOptions]=None) -> Radar:
        """Find all vehicles currently in a certain area."""
        if (not isinstance(rect, BoundingBox)):
            raise TypeError("argument rect: type BoundingBox expected")

        def generate() -> Async[Radar]:
            def bind(_arg1: Radar) -> Async[Radar]:
                return singleton.Return(_arg1)

            return singleton.Bind(self.asyncClient.AsyncRadar(rect, opt), bind)

        return await start_as_task(singleton.Delay(generate))

    async def tripsByName(self, name: str, opt: Optional[TripsByNameOptions]=None) -> TripsWithRealtimeData:
        """ Get all trips matching a name."""
        if (not isinstance(name, str)):
            raise TypeError("argument name: type string expected")

        if (opt is not None and not isinstance(opt, TripsByNameOptions)):
            raise TypeError("argument opt: type TripsByNameOptions expected")

        def generate() -> Async[TripsWithRealtimeData]:
            def bind(_arg1: TripsWithRealtimeData) -> Async[TripsWithRealtimeData]:
                return singleton.Return(_arg1)

            return singleton.Bind(self.asyncClient.AsyncTripsByName(name, opt), bind)

        return await start_as_task(singleton.Delay(generate))

    def productsOfMode(self, profile: Profile, mode: str) -> IndexMap_2[str, bool]:
        _checkProfileType(profile)

        return HafasAsyncClient_productsOfMode(profile, mode)

    def Dispose(self) -> None:
        __: HafasClient= self
        __.asyncClient.Dispose()
